README.markdown
jQuery *Long Click* Event
=========================
*Press & hold mouse/touch "long click" special event for [jQuery 1.4.x][jquery]*.
Inspired by similar events on touch-sensitive devices.


Using *Long Click*:
-------------------

### `.click( duration, handler )`
* Simply supply `duration` to the well-known [`.click`][jquery-click] method and you have a *long click*.
* This method is a shortcut for `.bind("longclick", handler)`.
* Returns *[jQuery][jquery-jquery]*.

### or `.longclick( [ duration ], [ handler ] )`
* If supplied, optional custom `duration` is used for target element(s).
* This method is a shortcut for `.click(duration, handler)` when at least `handler` is supplied and for `.trigger("longclick")` if called without arguments.
* Returns *[jQuery][jquery-jquery]*.


Parameter `duration`:
---------------------

* For how long (in milliseconds) mouse button must be pressed down (or touched) stationery to qualify as a *long click*.
* False value results in using the configured default.
* Default `duration` is **500** and is stored in `jQuery.longclick.duration` variable.


Event `"longclick"`:
--------------------

* Works with standard [`live`][jquery-live], [`bind`][jquery-bind], [`unbind`][jquery-unbind] and [`trigger`][jquery-trigger] event methods as one would expect.


Requirements:
-------------

1. Mighty [jQuery 1.4.x][jquery]:

    `<script src='http://code.jquery.com/jquery-latest.min.js' type='text/javascript'></script>`

2. and **[`jquery.longclick-min.js` (~1 kB)][download-min]**:

    `<script src='jquery.longclick-min.js' type='text/javascript'></script>`


License:
--------

Free for use in all personal or commercial projects under both [MIT][license-mit] and [GPL][license-gpl] licenses.


Fair Warning:
-------------

**Under intensive loving development**. v0.x works just fine, but is very "fluid" API-wise.
Till v1.0.

---
Copyright (c) 2010 [Petr Vostřel][pisi]
» Check out my other jQuery plugin [Reel][reel] for 360° image projection

[license-mit]: /pisi/Longclick/raw/master/MIT-LICENSE.txt
[license-gpl]: /pisi/Longclick/raw/master/GPL-LICENSE.txt
[jquery]: http://www.jquery.com/
[jquery-jquery]: http://docs.jquery.com/Types#jQuery
[jquery-live]: http://api.jquery.com/live/
[jquery-bind]: http://api.jquery.com/bind/
[jquery-unbind]: http://api.jquery.com/unbind/
[jquery-trigger]: http://api.jquery.com/trigger/
[jquery-click]: http://api.jquery.com/click/
[pisi]: http://petr.vostrel.cz/
[reel]: http://jquery.vostrel.cz/reel
[download-min]: http://github.com/pisi/Longclick/raw/master/jquery.longclick-min.js

CHANGELOG.markdown
Changelog
=========

0.3.3
-----
* Fixed GH-1 issue with dud "mouseout" binding.

0.3.2
-----
* Support for iPhone/iPad/iPod touch events has been added and confirmed using the iPhone Simulator.

0.3.1
-----
* Long click is dropped on "contextmenu" event to prevent right button from long clicking.

0.3
---
* Added `click` proxy. Ordinary built-in `click` can now be used to establish long click event simply by supplying a duration.
* The "mouseout" event is newly bound as one of annul events.
* A demonstration example has been added in `example.html`.
* Improved "click" event prevention in case of long click.
* (Un)binding and triggering tests added to test suite.
* Documentation comments & README improvements.

0.2
---
* `longclick` method now accepts an optional duration argument which overrides the default duration.
* `jQuery.longclick.delay` has been renamed to `jQuery.longclick.duration`.
* Invalid flag store key fixed.
* Documentation comments & README updates.
* Added QUnit test suite in `test/`

0.1
---
* Initial release.


test/tests.js
/**
 * jQuery Longclick Tests
 */
(function($){

  var
    $element

  module('', {
    setup: function setup(){
      $element= $('#element');
      event_fired= false;
    },
    teardown: function teardown(){
      $element.unbind('longclick');
      $('#live_sample').remove();
    }
  })

  test( 'Longclick event, method and configuration exists in the jQuery', function(){
    expect(5);

    /* Assert longclick event existence */
    equal( typeof $.event.special.longclick, 'object', 'Event');
    equal( typeof $.event.special.longclick.setup, 'function', 'Event setup');
    equal( typeof $.event.special.longclick.teardown, 'function', 'Event teardown');

    /* Asser longclick method and namespace */
    equal( typeof $.fn.longclick, 'function', '`longclick` method');
    equal( typeof $.longclick, 'object', 'Configuration');

  });
  test( 'Fire event by `longclick` method', function(){
    expect(1);

    /* Bind */
    $element.longclick(function record(){
      ok( this, '`longclick` event had fired')
    });

    /* Fire */
    $element.longclick();

  });
  test( 'Bind and fire event by `bind` and `trigger` methods', function(){
    expect(2);

    /* Bind */
    $element.bind('longclick', function record(){
      ok( this, '`longclick` event had fired')
    });
    ok( this, '`longclick` event has been bound')

    /* Fire */
    $element.trigger('longclick');

  })
  test( 'Unbind event by `unbind` method', function(){
    expect(1);

    /* Bind & unbind */
    $element.bind('longclick', function record(){
      ok( false, '`longclick` should not have had fired')
    }).unbind('longclick');

    /* Attempt to fire */
    $element.trigger('longclick');
    ok( this, '`Event unbound`');

  })
  test( 'Bind event by `live` method', function(){
    expect(1);

    /* Bind */
    $('#live_sample').live('longclick', function finish(){
      ok( this, 'Event successfully triggered');
    });

    /* Create element on the fly */
    $('<div id="live_sample"/>').appendTo($('body'));

    /* Trigger longclick event */
    $('#live_sample').longclick();

  });
  asyncTest( 'Fire after half a second by default', function(){
    expect(3);

    /* Mark start time */
    var
      started= Date.now()

    /* Assert configuration option */
    equal( typeof $.longclick.duration, 'number', 'Duration configuration');
    equal( $.longclick.duration, 500, 'Default duration value');

    /* Bind */
    $element.longclick(function finish(){
      ok( Math.abs(Date.now() - started - 500) < 50, 'In time (± 10 %)');
      /* The 10% tolerance is given because of an uncertainity based on client's CPU usage at the time */
      start();
    });

    /* Simulate long (endless) click */
    $element.trigger('mousedown');

  });
  asyncTest( 'Longclick `longclick` method accepts optional custom duration', function(){
    expect(1);

    /* Mark start time */
    var
      started= Date.now()

    /* Bind */
    $element.longclick(1000, function finish(){
      ok( Math.abs(Date.now() - started - 1000) < 50, 'In time (± 10 %)');
      /* The 10% tolerance is given because of an uncertainity based on client's CPU usage at the time */
      start();
    });

    /*
    Click the mouse button and hold until event is fired
    */
    $element.trigger('mousedown');
  });
  asyncTest( 'Longclick is dropped when mouse moves', function(){
    expect(1);

    /* Bind */
    $element.longclick(function finish(){ event_fired= true });

    /* Simulate long click interaction */
    $element.trigger('mousedown');
    setTimeout(function(){ $element.trigger('mousemove') }, 100);

    /* Check if has bee fired after a while */
    setTimeout(function(){
      equal( event_fired, false, 'Event had been dropped')
      start();
    }, 1000);

  });
  asyncTest( 'Longclick is dropped when mouse button is depressed', function(){
    expect(1);

    /* Bind */
    $element.longclick(function finish(){ event_fired= true });

    /* Simulate long click interaction */
    $element.trigger('mousedown');
    setTimeout(function(){ $element.trigger('mouseup') }, 100);

    /* Check if hasn't been fired after a while */
    setTimeout(function(){
      equal( event_fired, false, 'Event had been dropped')
      start();
    }, 1000);
  });
  asyncTest( 'Longclick is dropped on contextmenu event (right click)', function(){
    expect(1);

    /* Bind */
    $element.longclick(function finish(){ event_fired= true });

    /* Simulate long click interaction */
    $element.trigger('mousedown');
    setTimeout(function(){ $element.trigger('contextmenu') }, 100);

    /* Check if hasn't been fired after a while */
    setTimeout(function(){
      equal( event_fired, false, 'Event had been dropped')
      start();
    }, 1000);
  });
  asyncTest( 'Bind event by `click` method and supplying the duration', function(){
    expect(2);

    /* Bind
       - long click */
    $element.click(600, function record(){
      ok( this, '`longclick` event had fired')
      start();
    });

    /* - normal click */
    $element.click(function finish(){
      ok( this, 'Normal `click` still works')
    });

    /* Try normal click */
    $element.click();

    /* Simulate long click interaction */
    $element.trigger('mousedown');

  });
  asyncTest( 'Bind event by `longclick` method', function(){
    expect(2);

    /* Bind
       - using default duration */
    $element.longclick(function record(){
      ok( this, '`longclick` event had fired')
    });

    /* - using custom duration */
    $element.longclick(700, function record(){
      ok( this, 'Custom duration `longclick` event had fired')
      start();
    });

    /* Simulate long click interaction */
    $element.trigger('mousedown');

  });

})(jQuery);

jquery.longclick-min.js
/*
 Longclick Event
 Copyright (c) 2010 Petr Vostrel (http://petr.vostrel.cz/)
 Dual licensed under the MIT (MIT-LICENSE.txt)
 and GPL (GPL-LICENSE.txt) licenses.

 Version: 0.3.1
 Updated: 2010-06-04
*/
(function(a){function j(b){function d(){a(e).data(g,true);b.type=f;jQuery.event.handle.apply(e,k)}var e=this,k=arguments;a(this).data(g,false).data(i,setTimeout(d,a(this).data(h)||a.longclick.duration))}function l(){clearTimeout(a(this).data(i))}function m(b){if(a(this).data(g))return b.stopImmediatePropagation()||false}var n=a.fn.click;a.fn.click=function(b,d){if(!d)return n.apply(this,arguments);return a(this).data(h,b||null).bind(f,d)};a.fn.longclick=function(){var b=[].splice.call(arguments,0),
d=b.pop();b=b.pop();var e=a(this).data(h,b||null);return d?e.click(b,d):e.trigger(f)};a.longclick={duration:500};a.event.special.longclick={setup:function(){a(this).bind(o,j).bind([p,q,r,s].join(" "),l).bind(t,m)},teardown:function(){a(this).unbind(c)}};var f="longclick",c="."+f,o="mousedown"+c,t="click"+c,p="mousemove"+c,q="mouseup"+c,r="mouseout"+c,s="contextmenu"+c,h="duration"+c,i="timer"+c,g="fired"+c})(jQuery);


jquery.longclick.js
/**
 * jQuery Longclick Event
 * ======================
 * Press & hold mouse button "long click" special event for jQuery 1.4.x
 *
 * @license Longclick Event
 * Copyright (c) 2010 Petr Vostrel (http://petr.vostrel.cz/)
 * Dual licensed under the MIT (MIT-LICENSE.txt)
 * and GPL (GPL-LICENSE.txt) licenses.
 *
 * Version: 0.3.2
 * Updated: 2010-06-22
 */
(function($){

  /*
  `.click( duration, handler )`

  * Simply supply `duration` to the well-known `.click` method and you have a *long click*.
  * This method is a shortcut for `.bind("longclick", handler)`.
  * Returns *jQuery*.
  */
  var
    $_fn_click= $.fn.click

  $.fn.click= function click(duration, handler){
    /* Shortcircuit ordinary click calls */
    if (!handler) return $_fn_click.apply(this, arguments)
    /* Bind long click */
    return $(this).data(_duration_, duration || null).bind(type, handler)
  }

  /*
  `.longclick( [ duration ], [ handler ] )`

  * If supplied, optional custom `duration` is used for target element(s).
  * This method is a shortcut for `.click(duration, handler)` when at least `handler` is supplied
    and for `.trigger("longclick")` if called without arguments.
  * Returns *jQuery*.
  */
  $.fn.longclick= function longclick(){
    var
      args= [].splice.call(arguments, 0),
      handler= args.pop(),
      duration= args.pop(),
      $this= $(this).data(_duration_, duration || null)
    return handler ? $this.click(duration, handler) : $this.trigger(type)
  }

  /*
  Configuration
  */
  $.longclick= {
    /*
    * For how long (in milliseconds) mouse button must be pressed down (or touched) stationery
      to qualify as a *long click*.
    * False value results in using the configured default.
    * Default `duration` is **500** and is stored in `jQuery.longclick.duration` variable.
    */
    duration: 500
  }

  /*
  Bindings
  */
  $.event.special.longclick= {
    setup: function(data, namespaces){
      if (!(/iphone|ipad|ipod/i).test(navigator.userAgent)){
        /* normal technique for standard mouse-based interaction */
        $(this)
        .bind(_mousedown_, schedule)
        .bind([_mousemove_, _mouseup_, _mouseout_, _contextmenu_].join(' '), annul)
        .bind(_click_, click)
      }else{
        /* and special handling for touch-based interaction on iPhone-compatibile devices */
        touch_enabled(this)
        .bind(_touchstart_, schedule)
        .bind([_touchend_, _touchmove_, _touchcancel_].join(' '), annul)
        .bind(_click_, click)
        .css({ WebkitUserSelect: 'none' })
      }
    },
    teardown: function(namespaces){
      $(this).unbind(namespace)
    }
  }

  /*
  Commit subset of touch events to trigger jQuery events of same names
  */
  function touch_enabled(element){
    $.each('touchstart touchmove touchend touchcancel'.split(/ /), function bind(ix, it){
      element.addEventListener(it, function trigger_jquery_event(event){ $(element).trigger(it) }, false);
    });
    return $(element);
  }

  /*
  Handlers
  */
  function schedule(event){
    /* Check the timer isn't already running and drop if so */
    if ($(this).data(_timer_)) return;
    /* Catch in closure the `this` reference and `arguments` for later */
    var
      element= this,
      args= arguments
    /* Flag as "not fired" and schedule the trigger */
    $(this)
    .data(_fired_, false)
    .data(_timer_, setTimeout(scheduled, $(this).data(_duration_) || $.longclick.duration))

    function scheduled(){
      /* Flag as "fired" and rejoin the default event flow */
      $(element).data(_fired_, true)
      event.type= type
      jQuery.event.handle.apply(element, args)
    }
  }
  function annul(event){
    /* Annul the scheduled trigger */
    $(this).data(_timer_, clearTimeout($(this).data(_timer_)) || null)
  }
  function click(event){
    /* Prevent `click` event to be fired after button release once `longclick` was fired */
    if ($(this).data(_fired_)) return event.stopImmediatePropagation() || false
  }

  /*
  Frequent primitives and shortcuts
  */
  var
    type= 'longclick',
    namespace= '.' + type,

    /* Event strings */
    _mousedown_= 'mousedown'+namespace, _click_= 'click'+namespace,
    _mousemove_= 'mousemove'+namespace, _mouseup_= 'mouseup'+namespace,
    _mouseout_= 'mouseout'+namespace, _contextmenu_= 'contextmenu'+namespace,
    _touchstart_= 'touchstart'+namespace, _touchend_= 'touchend'+namespace,
    _touchmove_= 'touchmove'+namespace, _touchcancel_= 'touchcancel'+namespace,

    /* Storage keys */
    _duration_= 'duration'+namespace, _timer_= 'timer'+namespace, _fired_= 'fired'+namespace

})(jQuery);